### PR TITLE
Pure JIT for FnCAS on Linux.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi @mzhurovich, Cc @grixa,

So I hit the limits of our "JIT"-based optimization this week, and have invested into having a proper JIT.

After some half a day of research I realized LLVM and other similar tools are indeed too heavy, and wrote a lightweight one from scratch instead.

The code may look crazy, but if you begin reviewing it from the very `fncas/linux_native_jit/` directory and the test there, it's really just a few opcodes that are hardcoded to opcode-generating functions within an `{mmap()/mprotect()}`-invoking wrapper, and then a "thin" layer in the very FnCAS to compile the parsed expressions (an their gradients) into those opcodes. Per-commit review should be relatively easy as well.

Thanks!
Dima

PS: This is probably the largest piece of nontrivial code I wrote in 2.5 days in my entire life.

### CODE REVIEW

File: fncas/linux_native_jit/linux_native_jit.h [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+namespace linux_native_jit {
+
+// The signature is:
+// * Returns a `double`:              The value of the function.
+// * Uses the `double const* x`:      Input, the parameters vector `x[i]`.
+// * Uses the `double* o`:            Output, first the gradient, then the temporary memory buffer.
+// * Uses the `double (*f[])(double): External functions (`sin`, `exp`, etc.) to be called, to avoid dealing with PLT.
+typedef double (*pf_t)(double const* x, double* o, double (*f[])(double));
+
+constexpr static size_t const kLinuxNativeJITExecutableMageSize = 4096;
```
Comment (mzhurovich) [inline fncas/linux_native_jit/linux_native_jit.h]: Did you mean `PageSize`?

File: fncas/linux_native_jit/linux_native_jit.h [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+  CallableVectorUInt8(CallableVectorUInt8 const&) = delete;
+  CallableVectorUInt8(CallableVectorUInt8&) = delete;
+  CallableVectorUInt8(CallableVectorUInt8&&) = delete;
+  CallableVectorUInt8& operator=(CallableVectorUInt8 const&) = delete;
+  CallableVectorUInt8& operator=(CallableVectorUInt8&) = delete;
+  CallableVectorUInt8& operator=(CallableVectorUInt8&&) = delete;
+
+  double operator()(double const* x, double* o, double (*f[])(double)) const {
+    // HACK(dkorolev): Shift the buffers by 16 doubles (16 * 8 bytes) to have the load/save/etc. opcodes of same length.
+    // HACK(dkorolev): Shitf the functions buffer by one function (8 bytes) to even up the indirect call opcodes.
```
Comment (mzhurovich) [inline fncas/linux_native_jit/linux_native_jit.h]: `Shitf` -> `Shift`

File: fncas/linux_native_jit/linux_native_jit.h (hunk 1 above)
Comment (dkorolev) [reply fncas/linux_native_jit/linux_native_jit.h]: Yeah, I guess so. Thanks! https://github.com/C5T/Current/pull/840/commits/e0d2f4502367228d8b8cfb6a69437544b8c5a1c7

File: fncas/linux_native_jit/linux_native_jit.h (hunk 2 above)
Comment (dkorolev) [reply fncas/linux_native_jit/linux_native_jit.h]: :+1: https://github.com/C5T/Current/pull/840/commits/e0d2f4502367228d8b8cfb6a69437544b8c5a1c7

File: fncas/TODO.txt [hunk 3]
```diff
@@ -1,9 +1,19 @@
+TODO after adding real native Linux JIT:
+
+* Memory optimizations:
+  * Do a dry run to `.reserve()` just the right number of bytes to `mmap()`, and then generate the ultimate code just there.
+* Extra debug outputs.
+  * Time it took to differentiate.
+  * Size of the "binary" JIT code.
+* Extra optimizations:
+  * Actually operate on a `double*`, not a `vector<double>`, when the gradient is computed.
+  * Collapse constants.
```
Comment (mzhurovich) [inline fncas/TODO.txt:11]: Also `* Vectorized operations`?

**Comment by mzhurovich** [comment]
@dkorolev This is really cool, man :) I'll play with it on Mac later,
My $0.02:
* Let's add a reference to System V ABI - https://wiki.osdev.org/System_V_ABI
* More comments on what's exactly happening (stack/registers usage, assembler instructions brhind opcodes) won't hurt :)

**Comment by dkorolev** [comment]
Good idea. I think I'll add a separate `JIT.txt` file with the notes, and refer to it from where appropriate. Thanks!

**Comment by dkorolev** [comment]
@mzhurovich, thanks for the review, I've added a detailed [`JIT.md`](https://github.com/C5T/Current/blob/c89400545ac7057ac69c5814cbb5d278eea6abbc/fncas/JIT.md) in addition to fixing the typos.

PTAL!

Thanks,
Dima

**Comment by mzhurovich** [comment]
@dkorolev Yo! I've just successfully run Native JIT on my Mac by disabling Linux ifdef :) So I'd suggest to rename it from `linux_native_jit` to `x64_native_jit`.

File: fncas/linux_native_jit/benchmark.cc [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+DEFINE_int32(random_seed, 42, "The random seed to use. System- and compiler-dependent still.");
+DEFINE_int64(a, 1, "The left end of the range to uniformly draw the true variable values from.");
+DEFINE_int64(b, 9, "The [included] right end of the range to uniformly draw the true variable values from.");
+
+struct Data {
+  size_t const n;
+  size_t const m;
+  double const k;
+  size_t const q;                                  // Just `m * k` rounded to an integer.
+  std::vector<std::vector<size_t>> sparse_matrix;  // Sorted per training example.
```
Comment (mzhurovich) [inline fncas/linux_native_jit/benchmark.cc]: Let's stick to our convention of using explicitly-sized data types everywhere where we deal with serialization. Actually, it doesn't compile on my clang unless `size_t` is replaced with `uint64_t`.

File: fncas/JIT.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+* Low level in `linux_native_jit/linux_native_jit.h` and `linux_native_jit/test.cc`.
+* Mid level in `fncas/jit.h`.
+* Some higher level in `fncas/optimize.h` and the very `test.cc` from this directory.
+
+### Background
+
+At some point, the "interpreted" gradient computation code became the bottleneck, and compiling the autogenerated source code, be it via `as`, or `nasm`, or `clang`, to `dlopen()` the generated function later became too slow.
+
+I have done a bit of research around LLVM and simliar projects, quickly realized these is no simple and lightweight way to JIT-compile a mathematic expression into native machine opcodes, and hacked up one.
```
Comment (mzhurovich) [inline fncas/JIT.md]: `these is` -> `there is`

File: fncas/JIT.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+Generally, the generated assembly code is similar to what the already-existing assembly-generating logic does. All in all, it translates the AST tree of the mathematical expression of the function and its gradients directly into into the machine represention of the set of instructions that performs those computations.
```
Comment (mzhurovich) [inline fncas/JIT.md]: `into into` -> `into`

File: fncas/JIT.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+### Calling conventions
+
+Search for `opcodes::` in `linux_native_jit/test.cc` or in `fncas/jit.h` to quickly get to the very code generating pieces.
+
+The conventions are:
+* The function begins with `push rbx`, `mov rsi, rbx`, and and with `pop rbx`, `ret`.
+* The pointer to the input array of doubles is passed in `rdi`.
+* The pointer to the output (and intermediate) array of doubles is passed in `rsi` and stored in `rbx`.
+* The pointer to the array of external mathematical functions is stored in `rdx`.
+* When calling external functions, `rdi` and `rdi` are preserved on the stack.
```
Comment (mzhurovich) [inline fncas/JIT.md]: `rdi` and `rsi`?

File: fncas/linux_native_jit/benchmark.cc (hunk 4 above)
Comment (dkorolev) [reply fncas/linux_native_jit/benchmark.cc]: :+1: debc4015

File: fncas/JIT.md (hunk 5 above)
Comment (dkorolev) [reply fncas/JIT.md]: :+1: 8f40e9fb

File: fncas/JIT.md (hunk 7 above)
Comment (dkorolev) [reply fncas/JIT.md]: :+1: 8f40e9fb

File: fncas/JIT.md (hunk 6 above)
Comment (dkorolev) [reply fncas/JIT.md]: :+1: 8f40e9fb

**Comment by dkorolev** [comment]
Thanks @mzhurovich, especially for testing the code on a Mac!

Started Travis, fingers crossed.

Dima

**Comment by dkorolev** [comment]
And we are green!

**Pull request merged**